zrml/swaps/src/tests.rs

```diff
@@ -51,6 +51,28 @@ const _105: u128 = 105 * BASE;
 const _1234: u128 = 1234 * BASE;
 const _10000: u128 = 10000 * BASE;
 
+#[test_case(vec![ASSET_A, ASSET_A]; "case-1 with duplicate assets")]
+#[test_case(vec![ASSET_A, ASSET_B, ASSET_C, ASSET_D, ASSET_E, ASSET_A]; "case-2 with duplicate assets")]
+#[test_case(vec![ASSET_A, ASSET_B, ASSET_C, ASSET_D, ASSET_E, ASSET_E]; "case-3 with duplicate assets")]
+#[test_case(vec![ASSET_A, ASSET_B, ASSET_C, ASSET_A, ASSET_E, ASSET_D]; "case-4 with duplicate assets")]
```

Comment:
Not a big fan of the test names: `"case-{i} with duplicate assets"`. This means that the full test name is `pool_creation_fails_with_duplicate_assets::case_{i}_with_duplicate_assets`. I'd either go with detailed descriptions, or, if you want to go with running numbers, I'd just put `1`, `2`, etc.

```suggestion
#[test_case(vec![ASSET_A, ASSET_A]; "short vector")]
#[test_case(vec![ASSET_A, ASSET_B, ASSET_C, ASSET_D, ASSET_E, ASSET_A]; "start and end")]
#[test_case(vec![ASSET_A, ASSET_B, ASSET_C, ASSET_D, ASSET_E, ASSET_E]; "successive at end")]
#[test_case(vec![ASSET_A, ASSET_B, ASSET_C, ASSET_A, ASSET_E, ASSET_D]; "start and middle")]
```